src/Nethermind/Nethermind.Merge.AuRa/Shutter/Contracts/ValidatorRegistryContract.cs

```diff
@@ -0,0 +1,137 @@
+{
+    private const string getNumUpdates = "getNumUpdates";
+    private const string getUpdate = "getUpdate";
+    private readonly ulong _chainId = chainId;
+    private readonly ulong _messageVersion = messageVersion;
+
+    public UInt256 GetNumUpdates(BlockHeader blockHeader) => (UInt256)Call(blockHeader, getNumUpdates, Address.Zero, [])[0];
+
+    public Update GetUpdate(BlockHeader blockHeader, in UInt256 i)
+    {
+        var res = (ValueTuple<byte[], byte[]>)Call(blockHeader, getUpdate, Address.Zero, [i])[0];
+        Update update = new()
+        {
+            Message = res.Item1,
+            Signature = res.Item2
+        };
+        return update;
+    }
+
+    public bool IsRegistered(BlockHeader blockHeader, in Dictionary<ulong, byte[]> validatorsInfo, out HashSet<ulong> unregistered)
+    {
+        Dictionary<ulong, ulong?> nonces = [];
+        unregistered = [];
+        foreach (KeyValuePair<ulong, byte[]> validatorInfo in validatorsInfo)
+        {
+            nonces.Add(validatorInfo.Key, null);
+            unregistered.Add(validatorInfo.Key);
+        }
+
+        uint updates = (uint)GetNumUpdates(blockHeader);
+        for (uint i = 0; i < updates; i++)
+        {
+            Update update = GetUpdate(blockHeader, updates - i - 1);
+            Message msg = new(update.Message.AsSpan()[..46]);
+
+            // skip untracked validators
+            if (!validatorsInfo.ContainsKey(msg.ValidatorIndex))
+            {
+                continue;
+            }
+
+            if (msg.Version != _messageVersion)
+            {
+                if (logger.IsDebug) logger.Debug($"Registration message has wrong version ({msg.Version}) should be {_messageVersion}");
+                continue;
+            }
+
+            if (msg.ChainId != _chainId)
+            {
+                if (logger.IsDebug) logger.Debug($"Registration message has incorrect chain ID ({msg.ChainId}) should be {_chainId}");
+                continue;
+            }
+
+            if (!msg.ContractAddress.SequenceEqual(ContractAddress!.Bytes))
+            {
+                if (logger.IsDebug) logger.Debug($"Registration message contains an invalid contract address ({msg.ContractAddress.ToHexString()}) should be {ContractAddress}");
+                continue;
+            }
+
+            if (nonces[msg.ValidatorIndex].HasValue && msg.Nonce <= nonces[msg.ValidatorIndex])
+            {
+                if (logger.IsDebug) logger.Debug($"Registration message has incorrect nonce ({msg.Nonce}) should be {nonces[msg.ValidatorIndex]}");
+                continue;
+            }
+
+            if (!ShutterCrypto.CheckValidatorRegistrySignature(validatorsInfo[msg.ValidatorIndex], update.Signature, update.Message))
+            {
+                if (logger.IsDebug) logger.Debug("Registration message has invalid signature.");
+                continue;
+            }
+
+            // message is valid
+            nonces[msg.ValidatorIndex] = msg.Nonce;
```

Comment:
are validator indexes consecutive? If yes simple array would be better than Dictionary.

Reply:
Normally I think they would be, but not guaranteed